esp-hal-common/src/uart.rs

```diff
@@ -1442,67 +1484,66 @@ mod asynch {
-        /// (set by `set_rx_fifo_full_threshold`) and/or Rx AT_CMD character
-        /// interrupt if `set_at_cmd` was called.
+        /// Read async to buffer slice `buf`.
+        /// Waits until one of the following interrupts occurs:
+        /// - `RXFIFO_FULL`
+        /// - `RXFIFO_OVF`
+        /// - `AT_CMD_CHAR_DET` (only if `set_at_cmd` was called)
+        /// - `RXFIFO_TOUT` (only if `set_rx_timeout was called)
+        ///
+        /// The interrupts in question are enabled during the body of this
+        /// function. The method immediately returns when the interrupt
+        /// has already occured before calling this method (e.g. status
+        /// bit set, but interrupt not enabled)
         ///
         /// # Params
         /// - `buf` buffer slice to write the bytes into
         ///
         /// # Errors
         /// - `Err(RxFifoOvf)` when MCU Rx Fifo Overflow interrupt is triggered.
         ///   To avoid this error, call this function more often.
-        /// - `Err(Error::ReadNoConfig)` if neither `set_rx_fifo_full_threshold`
-        ///   or `set_at_cmd` was called
         ///
         /// # Ok
-        /// When succesfull, returns the number of bytes written to
-        /// buf
+        /// When succesfull, returns the number of bytes written to buf
         async fn read_async(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
             if buf.len() == 0 {
                 return Ok(0);
             }
 
-            let mut read_bytes = 0;
+            let mut interrupts_to_await = UartInterrupt::RXFIFO_FULL | UartInterrupt::RXFIFO_OVF;
 
             if self.at_cmd_config.is_some() {
-                if let Either3::Third(_) = select3(
-                    UartFuture::<T>::new(Event::RxCmdCharDetected),
-                    UartFuture::<T>::new(Event::RxFifoFull),
-                    UartFuture::<T>::new(Event::RxFifoOvf),
-                )
-                .await
-                {
-                    return Err(Error::RxFifoOvf);
-                }
-            } else {
-                if let Either::Second(_) = select(
-                    UartFuture::<T>::new(Event::RxFifoFull),
-                    UartFuture::<T>::new(Event::RxFifoOvf),
-                )
-                .await
-                {
-                    return Err(Error::RxFifoOvf);
-                }
+                interrupts_to_await |= UartInterrupt::AT_CMD_CHAR_DET;
             }
 
-            while let Ok(byte) = self.read_byte() {
-                if read_bytes < buf.len() {
-                    buf[read_bytes] = byte;
-                    read_bytes += 1;
-                } else {
-                    break;
-                }
+            // TODO: Abstract this
+            // This only works for esp32-c6 etc, because the timeout conf is not in the same
+            // register on all chips (esp32: UART_CONF1_REG)
+            if T::register_block()
```

Comment:
I would split this out into its own function where we can abstract over the differences. It's likely that the chips fall into 2 or 3 groups of "versions". _Usually_ the esp32/esp32s2 are similar, the esp32s3/esp32c3/esp32c2 are similar, and the final group is esp32c6/esp32h2.

Reply:
Yes. I will do this...